src/linear_algebra/tensor_algebra.lean

```diff
@@ -0,0 +1,239 @@
+  mul_one := by { rintros ⟨⟩, exact quot.sound rel.mul_one },
+  left_distrib := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, exact quot.sound rel.left_distrib },
+  right_distrib := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, exact quot.sound rel.right_distrib },
+  zero_mul := by { rintros ⟨⟩, exact quot.sound rel.zero_mul },
+  mul_zero := by { rintros ⟨⟩, exact quot.sound rel.mul_zero } }
+
+instance : inhabited (tensor_algebra R M) := ⟨0⟩
+
+instance : has_scalar R (tensor_algebra R M) :=
+{ smul := λ r a, quot.lift_on a (λ x, quot.mk _ $ ↑r * x) $
+  λ a b h, quot.sound (rel.mul_compat_right h) }
+
+instance : algebra R (tensor_algebra R M) :=
+{ to_fun := λ r, quot.mk _ r,
+  map_one' := rfl,
+  map_mul' := λ _ _, quot.sound rel.mul_scalar,
+  map_zero' := rfl,
+  map_add' := λ _ _, quot.sound rel.add_scalar,
+  commutes' := λ _, by { rintros ⟨⟩, exact quot.sound rel.central_scalar },
+  smul_def' := λ _ _, rfl }
+
+/--
+The canonical linear map `M →ₗ[R] tensor_algebra R M`.
+-/
+def ι : M →ₗ[R] (tensor_algebra R M) :=
+{ to_fun := λ m, quot.mk _ m,
+  map_add' := λ x y, quot.sound rel.add_lin,
+  map_smul' := λ r x, quot.sound rel.smul_lin }
+
+/--
+Given a linear map `f : M → A` where `A` is an `R`-algebra,
+`lift R M f` is the unique lift of `f` to a morphism of `R`-algebras
+`tensor_algebra R M → A`.
+-/
+def lift {A : Type u₃} [semiring A] [algebra R A] (f : M →ₗ[R] A) : tensor_algebra R M →ₐ[R] A :=
+{ to_fun := λ a, quot.lift_on a (lift_fun _ _ f) $ λ a b h,
+  begin
+    induction h,
+    { change f _ = f _ + f _, simp, },
+    { change f _ = (algebra_map _ _ _) * f _, rw linear_map.map_smul,
+      exact algebra.smul_def h_r (f h_a) },
+    { exact (algebra_map R A).map_add h_r h_s, },
+    { exact (algebra_map R A).map_mul h_r h_s },
+    { apply algebra.commutes },
+    { change _ + _ + _ = _ + (_ + _), rw add_assoc },
+    { change _ + _ = _ + _, rw add_comm, },
+    { change (algebra_map _ _ _) + lift_fun R M f _ = lift_fun R M f _, simp, },
+    { change _ * _ * _ = _ * (_ * _), rw mul_assoc },
+    { change (algebra_map _ _ _) * lift_fun R M f _ = lift_fun R M f _, simp, },
+    { change lift_fun R M f _ * (algebra_map _ _ _)= lift_fun R M f _, simp, },
+    { change _ * (_ + _) = _ * _ + _ * _, rw left_distrib, },
+    { change (_ + _) * _ = _ * _ + _ * _, rw right_distrib, },
+    repeat { set G := lift_fun R M f,
+      change G _ + G _ = G _ + G _ <|> change G _ * G _ = G _ * G _, rw h_ih, },
+    { change (algebra_map _ _ _) * _ = algebra_map _ _ _, simp },
+    { change _ * (algebra_map _ _ _) = algebra_map _ _ _, simp },
+  end,
+  map_one' := by change algebra_map _ _ _ = _; simp,
+  map_mul' := by { rintros ⟨⟩ ⟨⟩, refl },
+  map_zero' := by change algebra_map _ _ _ = _; simp,
+  map_add' := by { rintros ⟨⟩ ⟨⟩, refl },
+  commutes' := by tauto }
+
+variables {R M}
+
+@[simp]
+theorem ι_comp_lift {A : Type u₃} [semiring A] [algebra R A] (f : M →ₗ[R] A) :
+  (lift R M f).to_linear_map.comp (ι R M) = f := by {ext, refl}
```

Comment:
Is this similar theorem also useful?
```
@[simp]
theorem lift_comp_ι {A : Type u₃} [semiring A] [algebra R A] (g : tensor_algebra R M →ₐ[R] A) :
  lift R M (g.to_linear_map.comp (ι R M)) = g :=
begin
  symmetry,
  ext,
  rcases x,
  induction x,
  case pre.of: x { change (g.to_linear_map.comp (ι R M)) _ = _, refl },
  case pre.of_scalar: x { exact alg_hom.commutes g x },
  case pre.add: a b ha hb { change g (quot.mk _ _ + quot.mk _ _) = _,
    rw alg_hom.map_add,
    rw [ha, hb], refl },
  case pre.mul: a b ha hb { change g (quot.mk _ _ * quot.mk _ _) = _,
    rw alg_hom.map_mul,
    rw [ha, hb], refl },
end
```

With it, you can implement `lift_unique` as `⟨λ hyp, by simp [hyp], λ hyp, by simp [hyp]⟩